### PR TITLE
MONAI Models Integration

### LINKED ISSUE #211: Possible use of MONAI?
[This project](https://github.com/Project-MONAI/MONAI) might have relevant codebase that could be used within ivadomed, such as loss functions, transforms, architectures, etc.

### PR DESCRIPTION
**Pull request by uzaymacar**
#### GitHub

- [ ] I've given this PR a concise, self-descriptive, and meaningful title
- [ ] I've linked relevant issues in the PR body
- [ ] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [ ] I've assigned a reviewer

#### PR contents

- [ ] I've consulted [ivadomed's internal developer documentation](https://github.com/ivadomed/ivadomed/wiki) to ensure my contribution is in line with any relevant design decisions
- [ ] I've added [relevant tests](https://github.com/ivadomed/ivadomed/wiki/tests) for my contribution
- [ ] I've updated the [relevant documentation](https://github.com/ivadomed/ivadomed/wiki/documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description

As per [this comment](https://github.com/ivadomed/ivadomed/issues/211#issuecomment-1080964112), this PR explores interoperability between ivadomed and MONAI by adapting MONAI's `BasicUNet()` and `UNETR()` models to ivadomed. Specifically, the additions are:
* `monai` requirement added to the `requirements_common.txt` and model keywords added to `keywords.py`
*  `ivadomed_params_to_monai_params()` helper function to facilitate the mapping of ivadomed params to MONAI params in `models.py`
* Two model classes `MONAIUNet()` and `UNETR()` IN `models.py`

The `MONAIUNet()` model, for example, can now be used for training & testing by including the following in the config file:
```json
    "default_model": {
        "name": "MONAIUNet",
        "dropout_rate": 0.3,
        "features": [8, 16, 32, 32, 16, 8],
        "is_2d": false,
        "final_activation": "sigmoid",
        "length_3D": [128, 64, 32],
        "stride_3D": [128, 64, 32],
        "attention": false
    },
```

### TODO
* Benchmark MONAI models against ivadomed models using the [model_seg_ms_mp2rage](https://github.com/ivadomed/model_seg_ms_mp2rage) project.
* Implement `"attention": true` case as this is not covered in MONAI's `BasicUNet`; there is `AttentionUnet()` for this purpose in MONAI.
* Further generalize & beautify the implementation.
* Update documentation once/if we decide to merge this PR.

## Linked issues

Resolves #211.

### CODE REVIEW

File: ivadomed/models.py [hunk 1]
```diff
@@ -1440,6 +1441,79 @@ def forward(self, x):
...
+        self.mlp_dim = mlp_dim
+        self.num_heads = num_heads
+        self.pos_embed = pos_embed
+        self.dropout_rate = dropout_rate
+        self.final_activation = final_activation
+
+        ivadomed_params = {**self.__dict__, **kwargs}
+        monai_model_keys = ['img_size', 'feature_size', 'hidden_size', 'mlp_dim', 'num_heads', 'pos_embed']
+        monai_params = ivadomed_params_to_monai_params(ivadomed_params, monai_model_keys)
+        self.model = BasicUNet(**monai_params)
```
Comment (naga-karthik) [inline ivadomed/models.py:1498]: Shouldn't this be UNETR instead?

Comment (naga-karthik) [inline ivadomed/models.py:1498]: Shouldn't this be UNETR instead?

File: ivadomed/models.py [hunk 2]
```diff
@@ -6,6 +6,7 @@
 from torch.nn import init
 from pathlib import Path
 import torchvision.models
+from monai.networks.nets import BasicUNet
```
Comment (naga-karthik) [inline ivadomed/models.py:9]: Also need to import UNETR here

Comment (naga-karthik) [inline ivadomed/models.py:9]: Also need to import UNETR here

Comment (naga-karthik) [reply ivadomed/models.py]: Fixed in commit version https://github.com/ivadomed/ivadomed/pull/1116/commits/bfccd60402ab3b05b99b356d2c7c35f604441655

File: ivadomed/models.py [hunk 3]
```diff
@@ -1440,6 +1441,80 @@ def forward(self, x):
...
+
+    def forward(self, x):
+        out = self.model(x)
+
+        if hasattr(self, "final_activation") and self.final_activation not in ["softmax", "relu", "sigmoid"]:
+            raise ValueError("final_activation value has to be either softmax, relu, or sigmoid")
+        elif hasattr(self, "final_activation") and self.final_activation == "softmax":
+            out = self.softmax(out)
+        elif hasattr(self, "final_activation") and self.final_activation == "relu":
+            out = nn.ReLU()(out) / nn.ReLU()(out).max() if bool(nn.ReLU()(out).max()) else nn.ReLU()(out)
```
Comment (lifetheater57) [inline ivadomed/models.py:1470]: Why this division by the maximum value of the output? I know it was in ivadomed's Unet code, but what's the original motivation for it?

Comment (lifetheater57) [inline ivadomed/models.py:1470]: In this case, using the function API would be best.
```suggestion
            out = F.relu(out) / F.relu(out).max() if bool(F.relu(out).max()) else F.relu(out)
```

File: ivadomed/models.py [hunk 4]
```diff
@@ -1440,6 +1441,80 @@ def forward(self, x):
...
+    def forward(self, x):
+        out = self.model(x)
+
+        if hasattr(self, "final_activation") and self.final_activation not in ["softmax", "relu",
+                                                                               "sigmoid"]:
+            raise ValueError("final_activation value has to be either softmax, relu, or sigmoid")
+        elif hasattr(self, "final_activation") and self.final_activation == "softmax":
+            out = self.softmax(out)
+        elif hasattr(self, "final_activation") and self.final_activation == "relu":
+            out = nn.ReLU()(out) / nn.ReLU()(out).max() if bool(
```
Comment (lifetheater57) [inline ivadomed/models.py:1510]: Why this division by the maximum value of the output? I know it was in ivadomed's Unet code, but what's the original motivation for it?

Comment (lifetheater57) [inline ivadomed/models.py:1510]: In this case, using the function API would be best. (Remove next line if suggested code is accepted.)
```suggestion
            out = F.relu(out) / F.relu(out).max() if bool(
                F.relu(out).max()) else F.relu(out)
```

File: ivadomed/models.py [hunk 5]
```diff
@@ -1440,6 +1441,80 @@ def forward(self, x):
...
+        ivadomed_params = {**self.__dict__, **kwargs}
+        monai_model_keys = ['img_size', 'feature_size', 'hidden_size', 'mlp_dim', 'num_heads', 'pos_embed', 'norm_name']
+        monai_params = ivadomed_params_to_monai_params(ivadomed_params, monai_model_keys)
+        self.model = unetr.UNETR(**monai_params)
+
+    def forward(self, x):
+        out = self.model(x)
+
+        if hasattr(self, "final_activation") and self.final_activation not in ["softmax", "relu",
+                                                                               "sigmoid"]:
```
Comment (lifetheater57) [inline ivadomed/models.py:1505]: Why are there line returns like this in this model but not in `MONAIUNet`?

**Review by lifetheater57 — COMMENTED** [review]
Looks good in general. See comments in code for specific feedback.

File: ivadomed/models.py [hunk 6]
```diff
@@ -1520,3 +1595,30 @@ def get_model_filenames(folder_model):
         raise FileNotFoundError(folder_model)
 
     return str(fname_model), str(fname_model_metadata)
+
+
+def ivadomed_params_to_monai_params(ivadomed_params, monai_model_keys):
```
[the remark this answers is not in the document]
Comment (jcohenadad) [reply ivadomed/models.py:1600]: > Do we plan to integrate models from other frameworks in a foreseeable future?

no

File: ivadomed/models.py (hunk 3 above)
Comment (uzaymacar) [reply ivadomed/models.py:1470]: See [the docs on soft GT](https://ivadomed.org/configuration_file.html#soft-gt) for more context but essentially we use a "normalized" version of `relu` as the final activation to keep values in the range `0` and `1` (i.e., typical range for segmentation).

File: ivadomed/models.py (hunk 4 above)
Comment (uzaymacar) [reply ivadomed/models.py:1510]: Same answer as [here](https://github.com/ivadomed/ivadomed/pull/1116#discussion_r904173805).

File: ivadomed/models.py (hunk 3 above)
Comment (uzaymacar) [reply ivadomed/models.py:1470]: I agree! There are a couple of more `nn.ReLU()` calls to be replaced with `F.relu()` in the `models.py` in that case -- I will have a commit addressing this soon.

**Comment by lifetheater57** [comment]
@uzaymacar Did you add tests for these new models?
